crates/viewer/re_time_panel/src/data_density_graph.rs

```diff
@@ -442,6 +428,124 @@ pub fn data_density_graph_ui2(
+            },
+        );
+    }
+
+    // Small chunk heuristics:
+    // We want to render chunks as individual events, but it may be prohibitively expensive
+    // for larger chunks, or if the visible time range contains many chunks.
+    //
+    // We split a large chunk if:
+    // 1. The total number of events is less than some threshold
+    // 2. The number of events in the chunks is less than N, where:
+    //    N is relatively large for sorted chunks
+    //    N is much smaller for unsorted chunks
+
+    {
+        re_tracing::profile_scope!("add_data");
+
+        let can_render_individual_events =
+            config.max_total_chunk_events != 0 && total_events < config.max_total_chunk_events;
+
+        for (chunk, time_range, num_events_in_chunk) in chunk_ranges {
+            let should_render_individual_events = can_render_individual_events
+                && if chunk.is_time_sorted() {
+                    config.max_events_in_sorted_chunk != 0
+                        && num_events_in_chunk < config.max_events_in_sorted_chunk
+                } else {
+                    config.max_events_in_unsorted_chunk != 0
+                        && num_events_in_chunk < config.max_events_in_unsorted_chunk
+                };
+
+            if should_render_individual_events {
+                for (time, num_events) in chunk.num_events_cumulative_per_unique_time(&timeline) {
+                    data.add_chunk_point(time, num_events as usize);
+                }
+            } else {
+                data.add_chunk_range(time_range, num_events_in_chunk);
+            }
+        }
+    }
+
+    data
+}
+
+#[derive(Clone, Copy)]
+pub struct DensityGraphBuilderConfig {
+    pub max_total_chunk_events: usize,
+    pub max_events_in_sorted_chunk: usize,
+    pub max_events_in_unsorted_chunk: usize,
+}
+
+impl DensityGraphBuilderConfig {
+    /// All chunks will be rendered whole.
+    pub const NEVER_SPLIT_CHUNKS: Self = Self {
+        max_total_chunk_events: 0,
+        max_events_in_unsorted_chunk: 0,
+        max_events_in_sorted_chunk: 0,
+    };
+
+    /// All sorted chunks will be rendered as individual events,
+    /// and all unsorted chunks will be rendered whole.
+    pub const ALWAYS_SPLIT_SORTED_CHUNKS: Self = Self {
+        max_total_chunk_events: usize::MAX,
+        max_events_in_unsorted_chunk: 0,
+        max_events_in_sorted_chunk: usize::MAX,
+    };
+
+    /// All chunks will be rendered as individual events.
+    pub const ALWAYS_SPLIT_ALL_CHUNKS: Self = Self {
+        max_total_chunk_events: usize::MAX,
+        max_events_in_unsorted_chunk: usize::MAX,
+        max_events_in_sorted_chunk: usize::MAX,
+    };
+}
+
+impl Default for DensityGraphBuilderConfig {
+    fn default() -> Self {
+        Self {
+            // benchmarks report that this scales linearly,
+            // with ~1ms cost for a single chunk with 100k events
+            max_total_chunk_events: 100_000,
```

Comment:
This comment is confusing in this context.

And if it takes 1ms fof a single chunk, why is the limit 100k chunks?
What's more interesting here is the cost for an unsplit chunk, because that times `max_total_chunk_events` needs to be small enough.

---

Also: "1ms cost for a single chunk with 100k events" doing what? Viewing the individual events in it? If so, is it sorted or not?

Reply:
I tried to expand this as best as I could, but the limit is still somewhat arbitrary 🤷‍♂️